File: bridge/src/kiro-timeline-feed.ts
/**
 * Feeds an observed Kiro session's activity into the daemon's live timeline.
 *
 * #218 taught the Detail view to read Kiro's own transcript on demand, which
 * fixed the per-session query and nothing else: the main TIMELINE is not a
 * query, it is a STREAM that hook-driven agents push rows into as they work.
 * Kiro pushes nothing — no `kiro_*` hook has ever reached this daemon — so a
 * Kiro session appeared in the session list (HUD) while the timeline beside it
 * stayed empty. Answering "why is it in one and not the other" is answering
 * "one is pulled, the other is pushed".
 *
 * This is the missing producer. On every passive-observer refresh it reads
 * whatever is new in each Kiro session's transcript and emits it as ordinary
 * timeline rows, so the strip fills the way it does for every other agent.
 *
 * **A first sighting emits nothing.** It only records where the transcript
 * currently ends. Without that, every daemon start would dump each Kiro
 * session's entire history into a bounded activity log — evicting live rows
 * from other agents to re-tell a conversation from two days ago. Hook-driven
 * agents have the same property for free: their rows exist only from the
 * moment the daemon is listening. This makes Kiro behave the same way.
 */

import type { TimelineEntry } from '@agentdeck/shared';
import { kiroTimelineForSession } from './kiro-transcript-timeline.js';
import { debug } from './logger.js';

/** Rows read per session per tick. A Kiro turn is one prompt + one reply, so
 *  this is many turns' worth of headroom for a poll measured in seconds. */
const ROWS_PER_TICK = 8;

/** Watermarks kept for at most this many sessions, newest-first. A daemon that
 *  runs for weeks must not grow a map keyed by every session it ever saw. */
const MAX_TRACKED = 64;

export type KiroRowReader = (
  sessionId: string,
  opts: { since?: number; limit?: number },
) => TimelineEntry[];

export class KiroTimelineFeed {
  /** sessionId → newest row timestamp already accounted for. */
  private readonly watermark = new Map<string, number>();

  constructor(private readonly readRows: KiroRowReader = kiroTimelineForSession) {}

  /**
   * Rows to emit for this tick, oldest-first across all sessions.
   *
   * `sessionIds` is the CURRENT set of observed Kiro sessions; anything absent
   * from it is forgotten, so a finished session's watermark does not outlive it.
   */
  pump(sessionIds: string[]): TimelineEntry[] {
    const live = new Set(sessionIds);
    for (const id of [...this.watermark.keys()]) {
      if (!live.has(id)) this.watermark.delete(id);
    }

    const out: TimelineEntry[] = [];
    for (const id of sessionIds) {
      const since = this.watermark.get(id);
      let rows: TimelineEntry[];
      try {
        rows = this.readRows(id, { since, limit: ROWS_PER_TICK });
      } catch (err) {
        debug('daemon', `kiro-feed read failed for ${id}: ${String(err)}`);
        continue;
      }
      if (rows.length === 0) {
        // Seed an empty session at 0 so its first real row is emitted rather
        // than swallowed as "history".
        if (since === undefined) this.watermark.set(id, 0);
        continue;
      }
      const newest = rows[rows.length - 1].ts;
      if (since === undefined) {
        // First sighting: remember where the transcript ends, emit nothing.
        this.watermark.set(id, newest);
        debug('daemon', `kiro-feed seeded ${id} at ${newest}`);
        continue;
      }
      this.watermark.set(id, newest);
      out.push(...rows);
    }
    this.enforceCeiling();
    return out.sort((a, b) => (a.ts < b.ts ? -1 : a.ts > b.ts ? 1 : 0));
  }

  /** Tracked session count — for tests and diagnostics. */
  get size(): number {
    return this.watermark.size;
  }

  private enforceCeiling(): void {
    if (this.watermark.size <= MAX_TRACKED) return;
    // Oldest watermark first: the least recently active session is the one
    // least likely to produce the next row.
    const byAge = [...this.watermark.entries()].sort((a, b) => (a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0));
    for (const [id] of byAge) {
      if (this.watermark.size <= MAX_TRACKED) break;
      this.watermark.delete(id);
    }
  }
}

File: apple/AgentDeck/Daemon/Session/KiroTimelineFeed.swift
#if os(macOS)
// KiroTimelineFeed.swift — live timeline rows for observed Kiro sessions.
//
// Mirror of `KiroTimelineFeed` in bridge/src/kiro-timeline-feed.ts, and it
// exists for the same reason: the session list is PULLED (something scans and
// reports what it found) while the timeline is PUSHED (each agent's hooks add
// rows as it works). Kiro pushes nothing, so without a producer a Kiro session
// shows up in the HUD beside a timeline that stays empty forever — which is
// exactly what the user saw before this landed on the Node side.
//
// **A first sighting emits the last turn, and only that.** Replaying the whole
// transcript would push days of old conversation into a bounded activity log,
// evicting live rows from other agents. Emitting NOTHING was the first version
// and read as broken: a session that had been talking for an hour came back
// after a daemon restart with a visible session row and a blank strip beside
// it. One turn is enough to say what the session is doing.

import Foundation

@DaemonActor
final class KiroTimelineFeed {
    /// sessionId → newest row timestamp already accounted for.
    private var watermark: [String: Double] = [:]

    /// Watermarks kept for at most this many sessions. A daemon that runs for
    /// weeks must not grow a map keyed by every session it ever saw.
    private static let maxTracked = 64

    /// Rows to emit for this tick, oldest-first across all sessions.
    ///
    /// `sessionIds` is the CURRENT set of observed Kiro sessions (prefixed
    /// form); anything absent is forgotten, so a finished session's watermark
    /// does not outlive it.
    @discardableResult
    func pump(_ sessionIds: [String], now: Date = Date()) -> [DaemonTimelineEntry] {
        let live = Set(sessionIds)
        for id in watermark.keys where !live.contains(id) {
            watermark.removeValue(forKey: id)
        }

        let observed = LocalKiroObserver.observe(now: now)
        var byId: [String: LocalKiroObserver.Observed] = [:]
        for item in observed { byId["observed:kiro:\(item.sessionId)"] = item }

        var out: [DaemonTimelineEntry] = []
        for id in sessionIds {
            guard let session = byId[id] else { continue }
            let turns = LocalKiroObserver.readTurns(session.transcript)
            guard let newest = turns.last?.ts else {
                // Seed an empty session at 0 so its first real row is emitted
                // rather than swallowed as "history".
                if watermark[id] == nil { watermark[id] = 0 }
                continue
            }
            // Read the cut BEFORE moving the watermark: on a first sighting it
            // is the start of the last turn (so that turn, and nothing older,
            // is emitted); on every later tick it is what was already sent.
            let since = watermark[id] ?? Self.lastTurnStart(turns)
            watermark[id] = newest
            for turn in turns where turn.ts > since {
                out.append(DaemonTimelineEntry(
                    ts: turn.ts,
                    type: turn.isPrompt ? "chat_start" : "chat_response",
                    raw: Self.firstLine(turn.text),
                    detail: String(turn.text.prefix(4000)),
                    agentType: "kiro-cli",
                    projectName: session.projectName,
                    // The BARE uuid: timeline rows are keyed that way while the
                    // session list uses the prefixed form, and every client
                    // normalizes with ObservedAgentRules.rawSessionId before
                    // comparing the two.
                    sessionId: session.sessionId
                ))
            }
        }

        enforceCeiling()
        return out.sorted { $0.ts < $1.ts }
    }

    /// Tracked session count — for tests and diagnostics.
    var trackedCount: Int { watermark.count }

    private func enforceCeiling() {
        guard watermark.count > Self.maxTracked else { return }
        // Oldest watermark first: the least recently active session is the one
        // least likely to produce the next row.
        let byAge = watermark.sorted { $0.value < $1.value }
        for (id, _) in byAge {
            guard watermark.count > Self.maxTracked else { break }
            watermark.removeValue(forKey: id)
        }
    }

    /// A cut that admits the newest prompt and everything answering it.
    ///
    /// Walks back to the last prompt rather than taking a fixed count: a Kiro
    /// turn is one prompt plus however many reply records the agent wrote, so
    /// cutting by count would show a reply with no question above it. Returns
    /// just under that prompt's stamp, since the filter is strictly greater.
    private static func lastTurnStart(_ turns: [LocalKiroObserver.Turn]) -> Double {
        for turn in turns.reversed() where turn.isPrompt { return turn.ts - 1 }
        // No prompt in the window (a transcript whose tail is all replies):
        // fall back to the last two rows rather than the whole file.
        return turns.count > 2 ? turns[turns.count - 3].ts : 0
    }

    private static func firstLine(_ text: String, cap: Int = 120) -> String {
        let flat = text.split(whereSeparator: \.isNewline).joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard flat.count > cap else { return flat }
        return String(flat.prefix(cap - 1)) + "…"
    }
}
#endif
